ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/ChainDataPruner.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.chain;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.plugin.services.storage.KeyValueStorageTransaction;
+
+import java.util.Collection;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ChainDataPruner implements BlockAddedObserver {
+  private static final Logger LOG = LoggerFactory.getLogger(ChainDataPruner.class);
+  private final BlockchainStorage blockchainStorage;
+  private final ChainDataPrunerStorage prunerStorage;
+  private final long blocksToRetain;
+  private final long pruningFrequency;
+  private final ExecutorService pruningExecutor;
+  private static final int MAX_PRUNING_WORKER = 16;
+
+  public ChainDataPruner(
+      final BlockchainStorage blockchainStorage,
+      final ChainDataPrunerStorage prunerStorage,
+      final long blocksToRetain,
+      final long pruningFrequency) {
+    this.blockchainStorage = blockchainStorage;
+    this.prunerStorage = prunerStorage;
+    this.blocksToRetain = blocksToRetain;
+    this.pruningFrequency = pruningFrequency;
+    this.pruningExecutor =
+        new ThreadPoolExecutor(
+            1,
+            1,
+            60L,
+            TimeUnit.SECONDS,
+            new ArrayBlockingQueue<>(MAX_PRUNING_WORKER),
+            new ThreadPoolExecutor.DiscardPolicy());
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    LOG.debug("New block added event: " + event);
+    pruningExecutor.submit(
+        () -> {
+          // Get pruning mark
+          final long blockNumber = event.getBlock().getHeader().getNumber();
+          long pruningMark = prunerStorage.getPruningMark().orElse(blockNumber);
+          if (blockNumber < pruningMark) {
+            // Ignore and warn if block number < pruning mark, this normally indicates the
+            // blocksToRetain
+            // is too small.
+            LOG.warn(
+                "Block added event: "
+                    + event
+                    + " has a block number of "
+                    + blockNumber
+                    + " < pruning mark "
+                    + pruningMark);
+            return;
+          }
+          // Append block into fork blocks.
+          final KeyValueStorageTransaction tx = prunerStorage.startTransaction();
+          final Collection<Hash> forkBlocks = prunerStorage.getForkBlocks(blockNumber);
+          forkBlocks.add(event.getBlock().getHash());
+          prunerStorage.setForkBlocks(tx, blockNumber, forkBlocks);
+          // If a block is a new canonical head, start pruning.
+          if (event.isNewCanonicalHead()
+              && blockNumber - blocksToRetain - pruningMark >= pruningFrequency) {
```

Comment:
Finding this hard to comprehend. Can the subtractions be extracted into a variable and given a name?

Reply:
Updated. It should be clearer now